{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report Geoextent"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ideas for 'abstract'\n",
    "\n",
    "Academics studies generally include a geographical extent.\n",
    "\n",
    "Academic repositories store all kind of files that allow reasearchers to share information about their investigations. Some of those files include the analyzed datasets or the code used for their analysis. In this regard, the files available in each repositority include not only the information of individual measurements or methods but multiple information as the temporal or spatial extent of the studies. \n",
    "\n",
    "Temporal and spatial extent are properties that are present in almost all studies, however while searching for public repositories by an spatial component this query is most of the time limited to key words (e.g. name of countries). \n",
    "\n",
    "Multiple types of data -> hard to define a geographic extension\n",
    "\n",
    "\n",
    "Objective: \n",
    "\n",
    "Extract geographycal extent of academic repositories."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Database extraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These sets of functions download databases by using their DOI from four academic repositories. The functions also extracts the metadata of the database (e.g., location if available)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DOI to repository identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO\n",
    "# This functions use as an input the DOI of a dataset and returns the name of the academic repository where it\n",
    "# is hosted (Zenodo,Figshare,GFZ Data Services, Pangaea) and their ID in the corresponding repository."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Zenodo\n",
    "##### Getting list of records\n",
    "The following method extracts all the zenodo ids from an specific search term."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "def get_list_of_records(term,mb_size):\n",
    "    \n",
    "    response_hits = requests.get('https://zenodo.org/api/records/',\n",
    "                              params={'q':str(term),\n",
    "                                      \"access_right\":\"open\",\n",
    "                                      \"size\" : \"1\",\n",
    "                                     \"type\":\"dataset\"})\n",
    "    hits = response_hits.json()['hits']['total']\n",
    "    \n",
    "    print(\"{} repositories found for '{}' query search term\".format(hits,term))\n",
    "    zenodo_search = {}\n",
    "    \n",
    "    if hits > 0 :\n",
    "        \n",
    "        response = requests.get('https://zenodo.org/api/records/',\n",
    "                              params={'q':'geo',\"access_right\":\"open\",\"size\":str(hits),\"type\":\"dataset\"})\n",
    "        \n",
    "        content = response.json()\n",
    "\n",
    "    for i in range(0,hits):\n",
    "        \n",
    "        files = content['hits']['hits'][i]['files']\n",
    "        size = round(sum(f['size'] for f in files)/2 **20,1)\n",
    "        if size <= mb_size:\n",
    "            record_id = content['hits']['hits'][i]['conceptrecid']\n",
    "            doi = content['hits']['hits'][i]['doi']\n",
    "            title = content['hits']['hits'][i]['metadata']['title']\n",
    "            license = content['hits']['hits'][i]['metadata']['license']['id']\n",
    "            zenodo_search[record_id] = {\"doi\":doi,\"title\":title,\"license\":license,\"size_mb\":size}\n",
    "    print(\"{} out of {} repositories smaller than {} MB\".format(len(zenodo_search),hits,mb_size))\n",
    "\n",
    "    return zenodo_search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Figshare\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO\n",
    "\n",
    "# This function downloads a Figshare repository and extracts the available metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GFZ Data Services"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO\n",
    "\n",
    "# This function downloads a GFZ Data Services repository and extracts the available metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pangaea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO\n",
    "\n",
    "# This function downloads a Pangaea repository and extracts the available metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Geographic extent"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These sets of functions iterate through the repositories and extract the geospatial information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. If an extent is found (or if not = NA), Add record URL, ID, and some record metadata (names of contained files, author, repository URL, license, ...),  and the resulting extent in WKT in a local \"database\" in GeoPackage format.\n",
    "\n",
    "2. If not extent is found, store the record as visited and the list of filenames in a second data file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import geoextent.lib.extent as geoextent\n",
    "import zenodo_get as zget\n",
    "import tempfile\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def zenodo_geoextent(zenodo_id):\n",
    "    \n",
    "    with tempfile.TemporaryDirectory() as tmp_dir:\n",
    "        tmp_dir_dest = os.path.join(tmp_dir,str(zenodo_id))\n",
    "        zget.zenodo_get([str(zenodo_id),\"-e\",\"-o\",tmp_dir_dest])\n",
    "        f = geoextent.fromDirectory(tmp_dir_dest, True, True, True)\n",
    "        geo_extent = f\n",
    "        \n",
    "    return geo_extent"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Geoextent "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "216 repositories found for 'geo' query search term\n",
      "116 out of 216 repositories smaller than 100 MB\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Title: Whole-Genome and Epigenomic Landscapes of Etiologically Distinct Subtypes of Cholangiocarcinoma\n",
      "Keywords: \n",
      "Publication date: 2017-06-30\n",
      "DOI: 10.1158/2159-8290.CD-17-0368\n",
      "Total size: 0.1 MB\n",
      "\n",
      "Link: https://zenodo.org/api/files/501075a1-c50a-4b31-8146-f10fbe12cfb4/CCA_batch01_illumina_Gene_expression.xls   size: 0.1 MB\n",
      "\n",
      "Checksum is correct. (6817758b8cbf708c01b6be5e8a2db461)\n",
      "\n",
      "Link: https://zenodo.org/api/files/501075a1-c50a-4b31-8146-f10fbe12cfb4/CCA_batch02_illumina_Gene_expression.xls   size: 0.1 MB\n",
      "\n",
      "Checksum is correct. (8a84683f68af1a81320118dd75a6c025)\n",
      "All files have been downloaded.\n",
      "Title: A field-wide assessment of differential HT-seq reveals ubiquitous bias\n",
      "Keywords: ht-seq, rna-seq, p-values, gene expression omnibus\n",
      "Publication date: 2020-05-27\n",
      "DOI: 10.5281/zenodo.4046422\n",
      "Total size: 42.2 MB\n",
      "\n",
      "Link: https://zenodo.org/api/files/4639ab3d-74b4-4be0-8bdd-979d4d7edf92/geo-htseq-until-2019-12-31.tar.gz   size: 42.2 MB\n",
      "\n",
      "Checksum is correct. (acd145c3e55344e4867eded7460a7cad)\n",
      "All files have been downloaded.\n",
      "Title: Bicycle trips collected using Cyclists Geo-C geo-game\n",
      "Keywords: Bicycle, Cycling, Bike, Geo-game, Münster, Catelló, Malta, Geospatial, Coordinates, Urban Cycling, GIS\n",
      "Publication date: 2018-09-18\n",
      "DOI: 10.5281/zenodo.1421305\n",
      "Total size: 69.1 MB\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/bikepaths_segments.geojson   size: 2.2 MB\n",
      "\n",
      "Checksum is correct. (afb2e47de411eaaad22d955cf4ad0ccc)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/bikepaths_trips.geojson   size: 2.7 MB\n",
      "\n",
      "Checksum is correct. (b381ff0e95157443c543b7d4019cba50)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/frictions_cities.pdf   size: 4.5 MB\n",
      "\n",
      "Checksum is correct. (9cf0279ccb54434a718731d6a6fc669c)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/frictions.geojson   size: 0.2 MB\n",
      "\n",
      "Checksum is correct. (68aadada68f7106f6cfe2362cc9c5d30)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/location.geojson   size: 26.9 MB\n",
      "\n",
      "Checksum is correct. (1eb114cd0b40226f79f8e8e5b0563164)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/segments.geojson   size: 30.2 MB\n",
      "\n",
      "Checksum is correct. (51471327d43973068a201bea9622caa2)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/trips_od.geojson   size: 0.8 MB\n",
      "\n",
      "Checksum is correct. (1215fb9c911e440c6864ea7bac1ada4f)\n",
      "\n",
      "Link: https://zenodo.org/api/files/e4f76d83-435b-4230-8100-861554cb78f6/trips_tags.geojson   size: 1.7 MB\n",
      "\n",
      "Checksum is correct. (2b2bf5bc2b76eb6bd8e4395b344c2faa)\n",
      "All files have been downloaded.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There is no time value for GeoTIFF files\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Title: X-ray structure ensemble refinement of the second bromodomain of Pleckstrin homology domain interacting protein (PHIP) (space group P21212)\n",
      "Keywords: \n",
      "Publication date: 2020-10-29\n",
      "DOI: 10.5281/zenodo.4153881\n",
      "Total size: 9.6 MB\n",
      "\n",
      "Link: https://zenodo.org/api/files/bcd6dfe9-8217-46a9-a603-9e3aafdd16e1/mx8421v63_xPHIPAx1521_free.mtz   size: 1.3 MB\n",
      "\n",
      "Checksum is correct. (923f35e26c04f733924ad3bf21b18c00)\n",
      "\n",
      "Link: https://zenodo.org/api/files/bcd6dfe9-8217-46a9-a603-9e3aafdd16e1/PHIPA-P21212_ensemble_refinement_ensemble.geo   size: 1.6 MB\n",
      "\n",
      "Checksum is correct. (b1fc17409167b9317bef2bcc33819766)\n",
      "\n",
      "Link: https://zenodo.org/api/files/bcd6dfe9-8217-46a9-a603-9e3aafdd16e1/PHIPA-P21212_ensemble_refinement_ensemble.log   size: 0.5 MB\n",
      "\n",
      "Checksum is correct. (5e780aa40add4d3861b048b904866d20)\n",
      "\n",
      "Link: https://zenodo.org/api/files/bcd6dfe9-8217-46a9-a603-9e3aafdd16e1/PHIPA-P21212_ensemble_refinement_ensemble.mtz   size: 1.2 MB\n",
      "\n",
      "Checksum is correct. (f983d3856e0338a0a30c8b74cfef7e7b)\n",
      "\n",
      "Link: https://zenodo.org/api/files/bcd6dfe9-8217-46a9-a603-9e3aafdd16e1/PHIPA-P21212_ensemble_refinement_ensemble.pdb   size: 4.9 MB\n",
      "\n",
      "Checksum is correct. (362ecf88cab301f505f8de01c67fb2d8)\n",
      "\n",
      "Link: https://zenodo.org/api/files/bcd6dfe9-8217-46a9-a603-9e3aafdd16e1/refine9.pdb   size: 0.2 MB\n",
      "\n",
      "Checksum is correct. (d06b5a2fbe29ea0f87a3036bbbfc9bfd)\n",
      "All files have been downloaded.\n"
     ]
    }
   ],
   "source": [
    "#Get list of records\n",
    "geo  = get_list_of_records(\"geo\",mb_size=100)\n",
    "\n",
    "subset_list = list(geo.keys())[1:5]\n",
    "subset = {key: geo[key] for key in geo.keys() if key in subset_list}\n",
    "\n",
    "for i in subset:\n",
    "    if geoextent is not None:\n",
    "        subset[i]['geoextent'] = zenodo_geoextent(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "def rep_to_table(dict_geo):\n",
    "    \n",
    "    repository_id = []\n",
    "    title = []\n",
    "    doi = []\n",
    "    license = []\n",
    "    tbox = []\n",
    "    bbox = []\n",
    "    crs = []\n",
    "    \n",
    "    for i in dict_geo:\n",
    "\n",
    "        repository_info = dict_geo[i]\n",
    "        repository_id.append(i)\n",
    "        title.append(repository_info.get('title'))\n",
    "        doi.append(repository_info.get('doi'))\n",
    "        license.append(repository_info.get('license'))\n",
    "        geoextent = repository_info.get('geoextent')\n",
    "        tbox.append(geoextent.get('tbox'))\n",
    "        bbox.append(geoextent.get('bbox'))\n",
    "        crs.append(geoextent.get('crs'))\n",
    "        \n",
    "    d = {'repository_id': repository_id, 'title': title, 'doi':doi,'license':license,\n",
    "        'bbox':bbox,'tbox':tbox,'crs':crs}\n",
    "    \n",
    "    repositories = pd.DataFrame(d)\n",
    "    \n",
    "    return repositories\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repository_id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>license</th>\n",
       "      <th>bbox</th>\n",
       "      <th>tbox</th>\n",
       "      <th>crs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2632025</td>\n",
       "      <td>Whole-Genome and Epigenomic Landscapes of Etio...</td>\n",
       "      <td>10.1158/2159-8290.CD-17-0368</td>\n",
       "      <td>CC-BY-4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3747112</td>\n",
       "      <td>A field-wide assessment of differential HT-seq...</td>\n",
       "      <td>10.5281/zenodo.4046422</td>\n",
       "      <td>CC-BY-4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1421304</td>\n",
       "      <td>Bicycle trips collected using Cyclists Geo-C g...</td>\n",
       "      <td>10.5281/zenodo.1421305</td>\n",
       "      <td>CC-BY-4.0</td>\n",
       "      <td>[-99.09629, 0.0, 15.829299, 52.1808462]</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4153880</td>\n",
       "      <td>X-ray structure ensemble refinement of the sec...</td>\n",
       "      <td>10.5281/zenodo.4153881</td>\n",
       "      <td>CC-BY-4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  repository_id                                              title  \\\n",
       "0       2632025  Whole-Genome and Epigenomic Landscapes of Etio...   \n",
       "1       3747112  A field-wide assessment of differential HT-seq...   \n",
       "2       1421304  Bicycle trips collected using Cyclists Geo-C g...   \n",
       "3       4153880  X-ray structure ensemble refinement of the sec...   \n",
       "\n",
       "                            doi    license  \\\n",
       "0  10.1158/2159-8290.CD-17-0368  CC-BY-4.0   \n",
       "1        10.5281/zenodo.4046422  CC-BY-4.0   \n",
       "2        10.5281/zenodo.1421305  CC-BY-4.0   \n",
       "3        10.5281/zenodo.4153881  CC-BY-4.0   \n",
       "\n",
       "                                      bbox  tbox   crs  \n",
       "0                                     None  None  None  \n",
       "1                                     None  None  None  \n",
       "2  [-99.09629, 0.0, 15.829299, 52.1808462]  None  4326  \n",
       "3                                     None  None  None  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep_to_table(subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import itertools\n",
    "\n",
    "def extract_details(details,repository=1):\n",
    "    \n",
    "    filename = []\n",
    "    file_format = []\n",
    "    handler = []\n",
    "    bbox = []\n",
    "    tbox = []\n",
    "    crs = []\n",
    "    \n",
    "    for i in details:\n",
    "    \n",
    "        file = details[i]\n",
    "        \n",
    "        if file is None:\n",
    "            filename.append([i])\n",
    "            file_format_v = os.path.splitext(i)[1][1:]\n",
    "            if file_format_v is '':\n",
    "                file_format_v = 'undetected'\n",
    "            file_format.append([file_format_v])\n",
    "            handler.append([None])\n",
    "            bbox.append([None])\n",
    "            tbox.append([None])\n",
    "            crs.append([None])\n",
    "        else:\n",
    "            filename.append([i])\n",
    "            file_format.append([file.get('format')])    \n",
    "            handler_v = file.get('geoextent_handler')\n",
    "            bbox_v = file.get('bbox')\n",
    "            tbox_v = file.get('tbox')\n",
    "            crs_v = file.get('crs')\n",
    "            handler.append([handler_v])\n",
    "            bbox.append([bbox_v])\n",
    "            tbox.append([tbox_v])\n",
    "            crs.append([crs_v])\n",
    "            \n",
    "            if file.get('format')=='folder':\n",
    "                dictio = extract_details(file['details'])\n",
    "                filename.append(dictio['filename'])\n",
    "                file_format.append(dictio['format'])\n",
    "                handler.append(dictio['handler'])\n",
    "                bbox.append(dictio['bbox'])\n",
    "                tbox.append(dictio['tbox'])\n",
    "                crs.append(dictio['crs'])\n",
    "    \n",
    "    if any(isinstance(i, list) for i in filename):\n",
    "        filename = list(itertools.chain.from_iterable(filename))\n",
    "        file_format=  list(itertools.chain.from_iterable(file_format))\n",
    "        handler = list(itertools.chain.from_iterable(handler))\n",
    "        bbox= list(itertools.chain.from_iterable(bbox))\n",
    "        tbox = list(itertools.chain.from_iterable(tbox))\n",
    "        crs = list(itertools.chain.from_iterable(crs))\n",
    "        \n",
    "    d = {'filename': filename, 'format': file_format, 'handler':handler,'bbox':bbox,'tbox':tbox,'crs':crs}\n",
    "    \n",
    "    return(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def file_table(subset):\n",
    "    \n",
    "    filename = []\n",
    "    file_format = []\n",
    "    handler = []\n",
    "    tbox = []\n",
    "    bbox = []\n",
    "    crs = []\n",
    "    \n",
    "    for rep in subset:\n",
    "\n",
    "        file = subset[rep]['geoextent']['details']\n",
    "        repository = extract_details(file,1)\n",
    "\n",
    "        filename.append(repository.get('filename'))\n",
    "        file_format.append(repository.get('format'))\n",
    "        handler.append(repository.get('handler'))\n",
    "        tbox.append(repository.get('tbox'))\n",
    "        bbox.append(repository.get('bbox'))\n",
    "        crs.append(repository.get('crs'))\n",
    "\n",
    "    if any(isinstance(i, list) for i in filename):\n",
    "            \n",
    "        filename = list(itertools.chain.from_iterable(filename))\n",
    "        file_format=  list(itertools.chain.from_iterable(file_format))\n",
    "        handler = list(itertools.chain.from_iterable(handler))\n",
    "        bbox= list(itertools.chain.from_iterable(bbox))\n",
    "        tbox = list(itertools.chain.from_iterable(tbox))\n",
    "        crs = list(itertools.chain.from_iterable(crs))\n",
    "\n",
    "    d = {'filename': filename, 'format': file_format, 'handler':handler,'bbox':bbox,'tbox':tbox,'crs':crs}\n",
    "    files = pd.DataFrame(d)\n",
    "    return(files)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>format</th>\n",
       "      <th>handler</th>\n",
       "      <th>bbox</th>\n",
       "      <th>tbox</th>\n",
       "      <th>crs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCA_batch01_illumina_Gene_expression.xls</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>_rels</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.rels</td>\n",
       "      <td>undetected</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[Content_Types].xml</td>\n",
       "      <td>xml</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>theme</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>theme</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>theme1.xml</td>\n",
       "      <td>xml</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>_rels</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>themeManager.xml.rels</td>\n",
       "      <td>rels</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>themeManager.xml</td>\n",
       "      <td>xml</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CCA_batch02_illumina_Gene_expression.xls</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>_rels</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>.rels</td>\n",
       "      <td>undetected</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[Content_Types].xml</td>\n",
       "      <td>xml</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>theme</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>theme</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>theme1.xml</td>\n",
       "      <td>xml</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>_rels</td>\n",
       "      <td>folder</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>themeManager.xml.rels</td>\n",
       "      <td>rels</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>themeManager.xml</td>\n",
       "      <td>xml</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>md5sums.txt</td>\n",
       "      <td>txt</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>md5sums.txt</td>\n",
       "      <td>txt</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>geo-htseq-until-2019-12-31.tar.gz</td>\n",
       "      <td>gz</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>segments.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-0.3342129, 20.0001, 15.829299, 52.168266]</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>frictions_cities.pdf</td>\n",
       "      <td>pdf</td>\n",
       "      <td>handleRaster</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>bikepaths_segments.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-0.120553718143524, 35.88869779999997, 14.508...</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>bikepaths_trips.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-0.120553718143524, 35.88869779999997, 14.508...</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>md5sums.txt</td>\n",
       "      <td>txt</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>trips_od.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-99.09629, 0.0, 15.829298999999999, 52.167767]</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>location.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-0.33421290000000003, 0.0, 14.532264000000001...</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>frictions.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-0.069212307368285, 35.839033912250514, 14.50...</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>trips_tags.geojson</td>\n",
       "      <td>geojson</td>\n",
       "      <td>handleVector</td>\n",
       "      <td>[-99.09629, 19.457975, 15.829298999999999, 52....</td>\n",
       "      <td>None</td>\n",
       "      <td>4326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>refine9.pdb</td>\n",
       "      <td>pdb</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>PHIPA-P21212_ensemble_refinement_ensemble.log</td>\n",
       "      <td>log</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PHIPA-P21212_ensemble_refinement_ensemble.geo</td>\n",
       "      <td>geo</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>md5sums.txt</td>\n",
       "      <td>txt</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>PHIPA-P21212_ensemble_refinement_ensemble.mtz</td>\n",
       "      <td>mtz</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>mx8421v63_xPHIPAx1521_free.mtz</td>\n",
       "      <td>mtz</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>PHIPA-P21212_ensemble_refinement_ensemble.pdb</td>\n",
       "      <td>pdb</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         filename      format       handler  \\\n",
       "0        CCA_batch01_illumina_Gene_expression.xls      folder          None   \n",
       "1                                           _rels      folder          None   \n",
       "2                                           .rels  undetected          None   \n",
       "3                             [Content_Types].xml         xml          None   \n",
       "4                                           theme      folder          None   \n",
       "5                                           theme      folder          None   \n",
       "6                                      theme1.xml         xml          None   \n",
       "7                                           _rels      folder          None   \n",
       "8                           themeManager.xml.rels        rels          None   \n",
       "9                                themeManager.xml         xml          None   \n",
       "10       CCA_batch02_illumina_Gene_expression.xls      folder          None   \n",
       "11                                          _rels      folder          None   \n",
       "12                                          .rels  undetected          None   \n",
       "13                            [Content_Types].xml         xml          None   \n",
       "14                                          theme      folder          None   \n",
       "15                                          theme      folder          None   \n",
       "16                                     theme1.xml         xml          None   \n",
       "17                                          _rels      folder          None   \n",
       "18                          themeManager.xml.rels        rels          None   \n",
       "19                               themeManager.xml         xml          None   \n",
       "20                                    md5sums.txt         txt          None   \n",
       "21                                    md5sums.txt         txt          None   \n",
       "22              geo-htseq-until-2019-12-31.tar.gz          gz          None   \n",
       "23                               segments.geojson     geojson  handleVector   \n",
       "24                           frictions_cities.pdf         pdf  handleRaster   \n",
       "25                     bikepaths_segments.geojson     geojson  handleVector   \n",
       "26                        bikepaths_trips.geojson     geojson  handleVector   \n",
       "27                                    md5sums.txt         txt          None   \n",
       "28                               trips_od.geojson     geojson  handleVector   \n",
       "29                               location.geojson     geojson  handleVector   \n",
       "30                              frictions.geojson     geojson  handleVector   \n",
       "31                             trips_tags.geojson     geojson  handleVector   \n",
       "32                                    refine9.pdb         pdb          None   \n",
       "33  PHIPA-P21212_ensemble_refinement_ensemble.log         log          None   \n",
       "34  PHIPA-P21212_ensemble_refinement_ensemble.geo         geo          None   \n",
       "35                                    md5sums.txt         txt          None   \n",
       "36  PHIPA-P21212_ensemble_refinement_ensemble.mtz         mtz          None   \n",
       "37                 mx8421v63_xPHIPAx1521_free.mtz         mtz          None   \n",
       "38  PHIPA-P21212_ensemble_refinement_ensemble.pdb         pdb          None   \n",
       "\n",
       "                                                 bbox  tbox   crs  \n",
       "0                                                None  None  None  \n",
       "1                                                None  None  None  \n",
       "2                                                None  None  None  \n",
       "3                                                None  None  None  \n",
       "4                                                None  None  None  \n",
       "5                                                None  None  None  \n",
       "6                                                None  None  None  \n",
       "7                                                None  None  None  \n",
       "8                                                None  None  None  \n",
       "9                                                None  None  None  \n",
       "10                                               None  None  None  \n",
       "11                                               None  None  None  \n",
       "12                                               None  None  None  \n",
       "13                                               None  None  None  \n",
       "14                                               None  None  None  \n",
       "15                                               None  None  None  \n",
       "16                                               None  None  None  \n",
       "17                                               None  None  None  \n",
       "18                                               None  None  None  \n",
       "19                                               None  None  None  \n",
       "20                                               None  None  None  \n",
       "21                                               None  None  None  \n",
       "22                                               None  None  None  \n",
       "23        [-0.3342129, 20.0001, 15.829299, 52.168266]  None  4326  \n",
       "24                                               None  None  None  \n",
       "25  [-0.120553718143524, 35.88869779999997, 14.508...  None  4326  \n",
       "26  [-0.120553718143524, 35.88869779999997, 14.508...  None  4326  \n",
       "27                                               None  None  None  \n",
       "28    [-99.09629, 0.0, 15.829298999999999, 52.167767]  None  4326  \n",
       "29  [-0.33421290000000003, 0.0, 14.532264000000001...  None  4326  \n",
       "30  [-0.069212307368285, 35.839033912250514, 14.50...  None  4326  \n",
       "31  [-99.09629, 19.457975, 15.829298999999999, 52....  None  4326  \n",
       "32                                               None  None  None  \n",
       "33                                               None  None  None  \n",
       "34                                               None  None  None  \n",
       "35                                               None  None  None  \n",
       "36                                               None  None  None  \n",
       "37                                               None  None  None  \n",
       "38                                               None  None  None  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_table(subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GRAPH 1: Proportion of repositories with geospatial metadata**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NUMBER OF REPOSITORIES WITH GEOSPATIAL METADATA / NUMBER OF REPOSITORIES\n",
    "# what do they answer? \n",
    "# Proportion of repositories with geospatial metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GRAPH 2: Proportion of repositories with successful geospatial extraction (Geoextent)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NUMBER OF REPOSITORIES WITH SUCCESSFULL GEOSPATIAL EXTRACTION / NUMBER OF REPOSITORIES\n",
    "# what do they answer? \n",
    "# Does geoextent allow to extract geospatial information from repositories in a higher proportion that current metadata ?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GRAPH 3: Distribution of files in repositories with successful geospatial extraction (Geoextent)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISTRIBUTION OF FILES IN SUCCESSFUL GEOSPATIAL EXTRACTION\n",
    "# what do they answer?\n",
    "# What types of files (supported by geoextent) are more popular in the repositories ?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GRAPH 4: Distribution of (geo)files in repositories with unsuccessful geospatial extraction (Geoextent)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISTRIBUTION OF FILES IN UNSUCCESSFUL GEOSPATIAL EXTRACTIONS\n",
    "# what do they answer?\n",
    "# What type of files (geo but not supported by geoextent) are more popular in the repositories?"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
